Overhaul preset search/interaction type

jy2wong opened the pull request:
Future work:
* more intentional behaviour when hitting tab when no preset matches
* highlight input field text on click for easier typing

jy2wong commented:
Adds custom option, which is automatically selected if a user types in something with no search hits.
Switches back to single selection typeahead. Clicking anywhere pops up the entire list again, including the selected item.

Ready for code review.

blanchardjeremy commented:
I would make it so this info dialog is removed when they select the "build your own" option

![image](https://user-images.githubusercontent.com/32290/107579320-ca250180-6ba9-11eb-93cb-ba968fef6e50.png)

beshaya commented:
Also, if you have something selected, typing a new entry doesn't work until
you clear it or remove the text.

On Wed, Feb 10, 2021 at 2:11 PM Jeremy Blanchard <notifications@github.com>
wrote:

> I would make it so this info dialog is removed when they select the "build
> your own" option
>
> [image: image]
> <https://user-images.githubusercontent.com/32290/107579320-ca250180-6ba9-11eb-93cb-ba968fef6e50.png>
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/microcovid/microcovid/pull/740#issuecomment-777070027>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABM25LJHAXDVFKMV7TXHCLTS6MAAHANCNFSM4XMHVNSA>
> .
>

jy2wong commented:
> Also, if you have something selected, typing a new entry doesn't work until you clear it or remove the text.
> […](#)
> On Wed, Feb 10, 2021 at 2:11 PM Jeremy Blanchard ***@***.***> wrote: I would make it so this info dialog is removed when they select the "build your own" option [image: image] <https://user-images.githubusercontent.com/32290/107579320-ca250180-6ba9-11eb-93cb-ba968fef6e50.png> — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub <[#740 (comment)](https://github.com/microcovid/microcovid/pull/740#issuecomment-777070027)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABM25LJHAXDVFKMV7TXHCLTS6MAAHANCNFSM4XMHVNSA> .

Fixed locally, but waiting on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51176

jy2wong commented:
> I would make it so this info dialog is removed when they select the "build your own" option
> 
> ![image](https://user-images.githubusercontent.com/32290/107579320-ca250180-6ba9-11eb-93cb-ba968fef6e50.png)

done

jy2wong commented:
PTAL

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -37,6 +37,23 @@ export const SavedDataSelector: React.FunctionComponent<{
     props.scenarioNameSetter(key)
   }
 
+  const filterByCallback = (
```
Please add documentation or a descriptive function name.

jy2wong replied:
Done

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -48,18 +65,35 @@ export const SavedDataSelector: React.FunctionComponent<{
           }}
+          onFocus={() => {
+            if (savedDataRef.current) {
+              savedDataRef.current.getInput().select()
```
Leave a comment on this, please.

jy2wong replied:
Done

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -48,18 +65,35 @@ export const SavedDataSelector: React.FunctionComponent<{
+            shouldSelectHint: (shouldSelect: boolean, event) => {
+              if (
+                event.keyCode === 13 /* return */ ||
+                event.keyCode === 9 /* tab */
```
Is this working as intended? When I highlight an option and press tab, the dropdown closes and reverts to the previous selection.

jy2wong replied:
the event.keyCode === 13 part makes this work:

1. tab or click to focus input box
2. start typing indoor
3. the first entry that matches shows up as a hint
4. hitting enter selects the hint

It does seem like the keyCode === 9 part doesn't actually do anything. Will remove.

> When I highlight an option and press tab, the dropdown closes and reverts to the previous selection.

The sequence you describe doesn't work on the google search bar either and is also already present in the typeaheads on the current public-facing microcovid website.

jy2wong replied:
I have this fixed locally, but it requires another fix to the type definitions.

beshaya commented on src/data/prepopulated.ts:
```diff
@@ -255,4 +255,18 @@ export const prepopulated: {
     yourMask: 'basic',
     voice: 'silent',
   },
+  [i18n.t('scenario.custom')]: {
```
Is it necessary to create a dummy scenario here? Seems like an implementation detail of SavedDataSelector.tsx, could it be inserted there?

jy2wong replied:
I think so, actually. I'm not sure how else to reset the rest of the form. The comment is inaccurate and I'll remove it.

beshaya replied:
Oh, neat, the comment made me think it was no-op, but clearing the form is neat.

beshaya commented on src/locales/en.json:
```diff
@@ -214,6 +214,7 @@
     }
   },
   "scenario": {
+    "custom": "Build a custom scenario",
```
thoughts on "custom" vs "your own"?

jy2wong replied:
Done

beshaya reviewed (changes requested):
Looking good. Couple of requests for documentation, and possibly a bug with hitting 'tab'.

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -17,11 +18,10 @@ export const SavedDataSelector: React.FunctionComponent<{
   const prepopulatedOptions = Object.keys(prepopulated)
 
-  let selected: string[] = []
+  let selected2: string[] = []
```
why selected**2**?

jy2wong replied:
weird scope issues.

jy2wong replied:
I'll admit that my enthusiasm for switching it back and debugging is low.

jy2wong replied:
well, it at least seems to be weird scope issues with the *next* commit after this one. I'll switch it back for now.

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -48,19 +65,47 @@ export const SavedDataSelector: React.FunctionComponent<{
-          inputProps={{ autoComplete: 'chrome-off' }}
+          inputProps={{
+            autoComplete: 'chrome-off',
+            shouldSelectHint: (shouldSelect: boolean, event) => {
```
I'm still confused as to what this does. In the currently releases version, highlighting an option with the arrow keys and pressing return selects it. Same in the deploy preview. Could you leave a comment to explain what this does?

jy2wong replied:
Start typing "Outdoor" and press tab.

Reset, start typing "Outdoor" and press enter. Nothing happens on the current website.

jy2wong replied:
added a comment. Better?

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -48,19 +65,47 @@ export const SavedDataSelector: React.FunctionComponent<{
+            if (selected2.length === 0) {
+              selected2 = [i18n.t('scenario.custom')]
+              if (savedDataRef.current) {
+                savedDataRef.current.getInput().value = i18n.t(
```
I think this is unnecessary because setSavedData triggers a re-render, which will set the value via the `selected` prop

jy2wong replied:
Done and tested

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -48,19 +65,47 @@ export const SavedDataSelector: React.FunctionComponent<{
+                  'scenario.custom',
+                )
+              }
+              setSavedData(i18n.t('scenario.custom'))
```
Is this saying that, if I click into the typeahead and then change my mind, it will wipe out any data I've entered?

jy2wong replied:
if you tab through in any configuration (preset, modified from preset, custom blank, custom blank + some modifications), there is no change.

If you click/tab in, then click out, there is no change.

If you click/tab in, hit a few arrow keys, tab brings you out with no change. (same behaviour as current site; needs typescript definition update)

If you click/tab in, and hit any keyboard key other than tab/enter/arrows, it does clear the form on the way out. I suppose it's possible to add a check to see if the input bar text is empty, and if it's empty, don't reset.

jy2wong replied:
what this is *for* is 

I type in "hair cut" and it tells me I can build my own scenario. Enter already works, but tab did not: tab used to silently focus the next input element on the page. That means that if you had previously selected "one night stand" and then started typing hair cut, you would see a prompt to build your own scenario. If you then hit tab, you'd be in a weird spot where

* the precautions section is gone
* there's a risk level still showing for your "hair cut"
* you saw a suggestion to build your own and acted on it and have every reason to believe you are building your own hair cut scenario

jy2wong replied:
Here's a summary of the current and new behaviour in a table. Origin/Notes section is filled out for every change in behaviour (including internal behaviour that results in the same outward behaviour).

Do you still have concerns?
Would you be ok with deferring G for a future PR?

Setup | Bar autocompletes? | Current behaviour | New behaviour | Origin/Notes |
------ | ------------- | ---------------- | ---------------- | --- |
A. Type "outdoor", hit tab | Yes | Selects first entry | Selects first entry
B. Type "outdoor", hit enter | Yes | Nothing happens :x: | Selects first entry | shouldSelectHint
C. Type "masked", hit tab | No | Focus next element | Select custom scenario, reset form, focus next element | OnBlur
D. Type "masked", hit enter | No | Nothing | Nothing
E. Type "water", hit tab | No | Focus next element | Select custom scenario, reset form, focus next element | OnBlur
F. Type "water", hit enter | No | Nothing | Select custom scenario, reset form, focus next element | highlightOnlyResult=true
G. Focus bar, optionally type "masked", use arrow keys, hit tab | Yes | Keeps last selection :x: | Works. Uses new selection that's already highlighted and in the bar autocomplete hint. | Pending upstream change; not uploaded yet
H. Focus bar, optionally type "masked", use arrow keys, hit enter | Yes | Works | Works
I. No existing selection, focus in, optional highlight with mouse, focus out | | No selection change | No selection change | onBlur checks selected.length before acting
J. Existing selection, focus in, optional highlight with mouse, focus out | | No selection change | No selection change | onBlur checks selected.length before acting

blanchardjeremy commented:
I just wanna give infinity kudos and shout-outs to you, @jy2wong, for all the work you're doing on this!!

jy2wong commented:
Ok, I think this is ready for code review again. Tested with myself and one person who's seen and used the calculator before, but it could use further testing if anybody wants to use this preview instead of microcovid.org for the next couple scenarios.

Changes since last time:

* clicking on the magnifying glass pops up the typeahead menu
* successfully selecting a premade scenario drops you out of editing mode. The search bar will disappear. (Since "restart" selects the text inside the bar, starting a new scenario is still a one-click operation)
* new interaction type selector (removed dropdown; all options are visible at once)
* new "not-editing" informational mode with little icons that match the interaction type
* clear button to adjust interaction type in one click if you accidentally hit the wrong interaction type
* partially combined saved scenario selector and interaction type selector
* hide option to change interaction type for premades (only show for custom scenarios)
* hide remaining calculator features until either a premade or (custom + interaction type) are selected
* hide baseline transmission rate math behind expandable element

blanchardjeremy commented:
@jy2wong how do you feel about us closing this one?

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -42,20 +61,167 @@ const SavedDataLink: React.FunctionComponent<{
+  scenarioName: string
+  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void
+}> = (props): React.ReactElement => {
+  const { t } = useTranslation()
```
move to global scope?

jy2wong replied:
./src/components/calculator/SavedDataSelector.tsx
  Line 23:15:  React Hook "useTranslation" cannot be called at the top level. React Hooks must be called in a React function component or a custom React Hook function  react-hooks/rules-of-hooks

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -42,20 +61,167 @@ const SavedDataLink: React.FunctionComponent<{
+}> = (props): React.ReactElement => {
+  const iconFromInteractionType = (interaction: string) => {
+    const iconProps = { size: 24 }
+    if (props.interaction === 'oneTime') {
```
why is this props.interaction in some lines and interaction in others?

Also, consider a switch block instead?

jy2wong replied:
Done

beshaya commented on src/components/calculator/SavedDataSelector.tsx:
```diff
@@ -42,20 +61,167 @@ const SavedDataLink: React.FunctionComponent<{
+  setEditInteractionType: (newEdit: boolean) => void
+  interactionType: string
+}> = (props): React.ReactElement => {
+  const { t } = useTranslation()
```
Make a file constant

jy2wong replied:
ditto

beshaya commented on src/components/calculator/controls/VerticalControl.tsx:
```diff
@@ -0,0 +1,96 @@
+import { recordCalculatorOptionSelected } from 'components/Analytics'
+import { BaseFormValue } from 'data/data'
+
+export const VerticalControl: React.FunctionComponent<{
```
Could this be more descriptive? I don't have a sense of how this is supposed to look based on the component name. Alternatively, add documentation?

jy2wong replied:
done

beshaya commented on src/components/calculator/selectors/InteractionTypeSelector.tsx:
```diff
@@ -0,0 +1,160 @@
+                  ),
+                }}
+              >
+                Blah <strong>blah</strong>
```
Replace Blah with the gist of what will be rendered.

jy2wong replied:
Done

beshaya commented on src/locales/en.json:
```diff
@@ -143,7 +143,12 @@
+    "scenario_show_results": "Showing results for",
+    "scenario_guide_through": "Guiding you through",
+    "select_scenario_start_over": "Restart",
+    "saved_scenario_loaded_message": "Adjust the values below to match your situation and see how you can reduce risk (<4>based on research</4>!)",
```
"based on research" feels a little awkward to me. I don't insist.

beshaya commented on src/locales/en.json:
```diff
@@ -169,16 +174,18 @@
       "whats_this": "We’ve constructed a calculator that lets you estimate the risk of getting COVID from an activity or relationship in your daily life, using the <1>best research available</1>. We hope this tool will help hone your intuition, lower your stress levels, and figure out good harm-reduction strategies.</p><p>The first step in protecting your family and community is to avoid getting  ...
       "see_video": "If you'd like to watch a video walkthrough of how to use the calculator, click <1>here</1>.",
-      "call_to_action": "Calculate the approximate COVID risk of an activity or relationship"
+      "call_to_action": "Calculate approximate COVID risk"
```
COVID risk -> risk of contracting COVID?

jy2wong replied:
Done

beshaya commented on src/locales/en.json:
```diff
@@ -143,7 +143,12 @@
+    "saved_scenario_continue_custom": "Build your own scenario",
+    "scenario_show_results": "Showing results for",
+    "scenario_guide_through": "Guiding you through",
+    "select_scenario_start_over": "Restart",
```
Could this be "Choose a different scenario"?

jy2wong replied:
Done

jy2wong replied:
I'd prefer a shorter string here that people can pattern match from other contexts without having to fully read the prompt I think.

beshaya commented on src/locales/en.json:
```diff
@@ -143,7 +143,12 @@
+    "no_prebuilt_scenario_found": "No premade scenario found, but you may",
+    "saved_scenario_continue_custom": "Build your own scenario",
+    "scenario_show_results": "Showing results for",
+    "scenario_guide_through": "Guiding you through",
```
I find this to be a bit over-selling. I'd think of guiding you through as having things pop up as you go along with detailed instructions.

jy2wong replied:
"Gathering more information about"?

jy2wong replied:
"Add details about"?

jy2wong replied:
ehhh. I just removed it altogether.

beshaya commented on src/locales/en.json:
```diff
@@ -143,7 +143,12 @@
     "prevalence_info_source_information": "Prevalence data consolidated from <1>Johns Hopkins CSSE</1> (reported cases), <4>Covid Act Now</4> (US positive test rates), <7>Our World in Data</7> (international positive test rates), and <10>Coronavirus COVID-19 România</10> (Romania reported cases).",
     "select_scenario": "Start with a premade scenario or <2>build your own</2>:",
-    "no_prebuilt_scenario_found": "We don't have this scenario yet, but you can build your own below.",
+    "no_prebuilt_scenario_found": "No premade scenario found, but you may",
```
These two need to be a single entry; other languages may not have the same clause order.

jy2wong replied:
removed the "but you may".

beshaya commented on src/pages/Calculator.tsx:
```diff
@@ -67,6 +68,19 @@ export const Calculator = (): React.ReactElement => {
 
+  const { t } = useTranslation()
+
+  const [editScenario, setEditScenario] = useState(
```
Having the logic for what is visible be separated from the data makes it hard to reason about.

It looks to me that both editScenario and editInteractionType are (approximately) equal to "is a preset selected"

Could we add the selected preset to CalculatorData and use this directly? It would also cause share links to show the same scenario.

jy2wong replied:
editInteractionType is "does the user need to set oneTime, workplace, repeated or partner right now?" which is strongly tied to "is custom" but not equivalent.

editScenario is "no preset nor custom"

editScenario seems like it could be removed by moving scenarioName/setScenarioName to CalculatorData. I'll look into it.

jy2wong replied:
> Could we add the selected preset to CalculatorData and use this directly? It would also cause share links to show the same scenario.

Done. Also cleaned out the edit{InteractionType,Scenario} from Calculator.tsx.

beshaya reviewed (commented):
Two high level comments:

* Getting to the custom scenario seems a little clunky to me. I'd expect the "restart" button to clear the screen and put me in a custom scenario. Since it focuses the preset selector, maybe "custom" should be the first entry in that list?
* Should "going to work" be a preset instead of an interaction type? At the least, I'd have expected it to set the duration to 40 hours.

jy2wong commented:
> Two high level comments:
> 
> * Getting to the custom scenario seems a little clunky to me. I'd expect the "restart" button to clear the screen and put me in a custom scenario. Since it focuses the preset selector, maybe "custom" should be the first entry in that list?

When I see the word restart, I think "take me back to the beginning". I think going straight to a custom scenario is a bit of a conceptual leap. I remember flipping through a lot of feedback asking for "a way to enter another scenario" (i.e. feedback that the Reset button wasn't discoverable enough or located where people wanted to find it).

(also the "build your own" link goes straight to custom scenario. Does changing the text to "build a custom scenario" achieve the same goal?)

> * Should "going to work" be a preset instead of an interaction type? At the least, I'd have expected it to set the duration to 40 hours.

Work situation definitely needs improvement, but I think it's out of scope for this PR.

beshaya commented:
> When I see the word restart, I think "take me back to the beginning". I think going straight to a custom scenario is a bit of a conceptual leap. I remember flipping through a lot of feedback asking for "a way to enter another scenario" (i.e. feedback that the Reset button wasn't discoverable enough or located where people wanted to find it).

Yeah, thinking about it more, I think "custom scenario" in my head is the default because clearing all data before put you in what is a custom scenario now. That's not necessarily ideal.

> > * Should "going to work" be a preset instead of an interaction type? At the least, I'd have expected it to set the duration to 40 hours.
> 
> Work situation definitely needs improvement, but I think it's out of scope for this PR. Defer?

Deferring seems right to me.

Overall this seems like an improvement.

beshaya commented on src/components/calculator/selectors/InteractionTypeSelector.tsx:
```diff
@@ -0,0 +1,154 @@
+}> = (props): React.ReactElement => {
+  const { t } = useTranslation()
+  const [showInteractionTypeMath, setShowInteractionTypeMath] = useState(false)
+  const getInteractionRiskMessage = (interactionType: string) => {
```
Nit, would be cleaner as case/switch.

jy2wong replied:
done

beshaya commented on src/components/calculator/selectors/InteractionTypeSelector.tsx:
```diff
@@ -0,0 +1,154 @@
+        {t(Interaction[props.interactionType].label)}
+      </Dropdown.Toggle>
+      <Dropdown.Menu className="w-100">
+        {Object.keys(Interaction).map((value, index) => {
```
Consider `Object.entries(Interaction).map(([key, value], index)` instead.

jy2wong replied:
done

beshaya commented on src/pages/Calculator.tsx:
```diff
@@ -98,11 +101,14 @@ export const Calculator = (): React.ReactElement => {
     addAlert('Link copied to clipboard!')
   }
 
+  const [pointsVisible, setPointsVisible] = useState(false)
```
Is this used outside of the useMemo block below? It looks to me like `pointsVisible = points !== -1`

jy2wong replied:
removed

beshaya commented on src/pages/Calculator.tsx:
```diff
@@ -146,6 +153,13 @@ export const Calculator = (): React.ReactElement => {
     calculatorData.interaction,
   )
 
+  const editScenario =
```
nit: could this be scenarioSelected or scenarioEditable?

jy2wong replied:
done

beshaya commented on src/pages/Calculator.tsx:
```diff
@@ -146,6 +153,13 @@ export const Calculator = (): React.ReactElement => {
+  const editScenario =
+    calculatorData.scenarioName === undefined ||
+    calculatorData.scenarioName === ''
+  const editInteractionType =
```
nit: could this be interactionSelected or interactionEditable?

jy2wong replied:
done

beshaya commented on src/pages/Calculator.tsx:
```diff
@@ -272,124 +260,151 @@ export const Calculator = (): React.ReactElement => {
-                        vaccineDoses: calculatorData.yourVaccineDoses,
-                        vaccineType: calculatorData.yourVaccineType,
+                      setSavedData={(newScenario: string): void => {
+                        let foundData:
```
Why is this type so complicated? prepopulated is [`key: string]: PartialData`, so shouldn't foundData be of type `PartialData?`

jy2wong replied:
I copied it over from the existing `foundData`. `null` still needs to be there in case of invalid scenario names like `''` but yeah, it looks like it can't actually be a `calculatorData`.

beshaya reviewed (approved):
LGTM % some nits.

Thanks for sticking it out on this one!

blanchardjeremy commented:
You're my hero, @jy2wong! This is such a great improvement! :)
